File: raspberry-pi-ios-client/RaspberryPiClient/Presentation/Service/Joystick/KnobPosition.swift
//
//  KnobPosition.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/12/29.
//

import SwiftUI

struct KnobPosition {
    private var position: CGSize
    private static let screenWidth = UIScreen.main.bounds.width
    private static var maxDistance: CGFloat {
        parentCircleWidth / 2 - knobCircleWidth / 2
    }

    /// Waiting factor
    private let WF: Double = 20.0

    static var chevronWidth: CGFloat {
        screenWidth * 0.875
    }
    static var parentCircleWidth: CGFloat {
        screenWidth * 0.625
    }
    static var shadowCircleWidth: CGFloat {
        screenWidth * 0.375
    }
    static var knobCircleWidth: CGFloat {
        screenWidth * 0.25
    }

    init(width: CGFloat, height: CGFloat) {
        self.position = CGSize(width: width, height: height)
    }

    // MARK: Range -2.0 ~ +2.0 to use ROS2 service

    var x: Float {
        let normalizedValue = (position.width / Self.maxDistance) * WF
        let clampedValue = max(-10.0, min(10.0, normalizedValue))
        return Float(floor(clampedValue * 100) / 100)
    }

    var y: Float {
        let normalizedValue = (position.height / Self.maxDistance) * WF
        let clampedValue = max(-10.0, min(10.0, normalizedValue))
        return Float(floor(clampedValue * 100) / 100)
    }

    // MARK: Range -maxDistance ~ +maxDistance to display Joystic knob

    var xCGFloat: CGFloat {
        position.width
    }

    var yCGFloat: CGFloat {
        position.height
    }

    var theta: Float {
        let radians = atan2(x, y)
        let degrees = radians * 180.0 / .pi
        return degrees >= 0 ? degrees : degrees + 360.0
    }

    var radian: Float {
        -atan2(x, y) * Float(WF / 2.0)
    }

    mutating func changePosition(_ value: DragGesture.Value) {
        // Invert the sign of the Y coordinate to make the upward direction positive.
        // Because the downward direction of the screen is positive in SwiftUI's coordinate system.
        let normalizedY = -value.translation.height
        let distance = hypot(value.translation.width, normalizedY)

        // Knob position
        if distance > Self.maxDistance {
            let scale = Self.maxDistance / distance
            position = CGSize(
                width: value.translation.width * scale,
                height: normalizedY * scale
            )
        } else {
            position = CGSize(
                width: value.translation.width,
                height: normalizedY
            )
        }
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_android_client/presentation/screen/service/compose/KnobPosition.kt
package com.shusuke.raspberry_pi_android_client.presentation.screen.service.compose

import kotlin.math.atan2
import kotlin.math.floor
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min

/**
 * ジョイスティックノブ位置（iOS [KnobPosition] に相当）。
 *
 * @param screenWidthPx ジョイスティック配置画面の幅（px）。Compose 側から渡す。
 */
data class KnobPosition(
    val width: Float = 0f,
    val height: Float = 0f,
    val screenWidthPx: Float = 1080f,
) {
    companion object {
        private const val WF = 20f

        fun parentCircleWidth(screenWidthPx: Float): Float = screenWidthPx * 0.625f

        fun shadowCircleWidth(screenWidthPx: Float): Float = screenWidthPx * 0.375f

        fun knobCircleWidth(screenWidthPx: Float): Float = screenWidthPx * 0.25f

        fun chevronWidth(screenWidthPx: Float): Float = screenWidthPx * 0.875f
    }

    private val maxDistance: Float
        get() = parentCircleWidth(screenWidthPx) / 2f - knobCircleWidth(screenWidthPx) / 2f

    val x: Float
        get() {
            val normalized = (width / maxDistance) * WF
            val clamped = max(-10.0, min(10.0, normalized.toDouble()))
            return (floor(clamped * 100.0) / 100.0).toFloat()
        }

    val y: Float
        get() {
            val normalized = (height / maxDistance) * WF
            val clamped = max(-10.0, min(10.0, normalized.toDouble()))
            return (floor(clamped * 100.0) / 100.0).toFloat()
        }

    val theta: Float
        get() {
            val radians = atan2(x.toDouble(), y.toDouble())
            val degrees = radians * 180.0 / Math.PI
            return (if (degrees >= 0) degrees else degrees + 360.0).toFloat()
        }

    val radian: Float
        get() = -atan2(x.toDouble(), y.toDouble()).toFloat() * (WF / 2f)

    fun withDragTranslation(translationX: Float, translationY: Float): KnobPosition {
        val normalizedY = -translationY
        val distance = hypot(translationX.toDouble(), normalizedY.toDouble()).toFloat()
        if (distance > maxDistance) {
            val scale = maxDistance / distance
            return copy(width = translationX * scale, height = normalizedY * scale)
        }
        return copy(width = translationX, height = normalizedY)
    }
}
